src/sentry/processing/realtime_metrics/redis.py

```diff
@@ -164,22 +163,28 @@ def get_durations_for_project(self, project_id: int) -> Iterable[base.DurationHi
         This may throw an exception if there is some sort of issue fetching durations from the redis
         store.
         """
-        key_prefix = f"{self._histogram_key_prefix()}:{project_id}:"
-        keys = sorted(
-            self.cluster.scan_iter(
-                match=key_prefix + "*",
-            )
-        )
+        bucket_size = self._histogram_bucket_size
+        now = int(time.time())
+        now_timestamp = now - now % bucket_size
 
-        for key in keys:
-            _, timestamp_raw = key.split(key_prefix)
-            timestamp_bucket = int(timestamp_raw)
+        first_timestamp = int(now - self._histogram_ttl / datetime.timedelta(seconds=1))
+        first_timestamp = first_timestamp - first_timestamp % bucket_size
 
-            histogram_raw = self.cluster.hgetall(key)
-            histogram = base.BucketedDurations(
-                {int(duration): int(count) for duration, count in histogram_raw.items()}
+        timestamps = range(first_timestamp, now_timestamp + bucket_size, bucket_size)
+        durations = range(0, 600, 10)
+
+        for timestamp in timestamps:
+            histogram_redis = self.cluster.hgetall(
+                f"{self._histogram_key_prefix()}:{project_id}:{timestamp}"
+            )
```

Comment:
microoptimization: could we pipeline all of the `hgetall`s for all of the timestamps, so we get all of them in one round-trip to redis instead of fetching them one by one?

Reply:
I tried, but I can't figure out how.

Reply:
i gave this a shot in 32c043994ad5b87b1856dd17e60e994a2ef94557 - let me know if this was what you tried, and if there were some issues that i might've overlooked. i'm relying on the tests to fail if this approach is incorrect, so it's possible that i may have missed something.

Reply:
Nicely done, I didn't realize that `pipeline.execute()` returns a value.